Add a new grammar renderer

ehuss opened the pull request:
This introduces a new grammar renderer. Instead of trying to write the grammar in markdown/html hybrid, this introduces a new syntax that is parsed by the mdbook-spec plugin. This grammar is then converted into markdown/html hybrid, and also to railroad diagrams.

There are a lot of changes here (and some can be split into separate PRs if desired). A general overview of what to see here:

- Grammar rules are now written inside a code block (instead of a blockquote). The syntax is pretty similar to the old syntax with various small changes. See the `docs/grammar.md` file for a complete description.
- There is now a summary chapter which shows the entire grammar all on one page.
- The grammar is parsed by `mdbook-spec/src/grammar/parser.rs` into an internal representation.
- The internal representation is converted to markdown in `mdbook-spec/src/grammar/render_markdown.rs`, and railroad diagrams in `mdbook-spec/src/grammar/render_railroad.rs`.
    - The railroad diagrams are generated using the [railroad crate](https://crates.io/crates/railroad).
    - There is a toggle button the show/hide the railroad diagrams. It uses localstorage to keep that state sticky.
- The basic definitions and driver in the mdbook plugin is in `mdbook-spec/src/grammar.rs`. There are several pieces here:
    - The internal representation.
    - Code to load the grammar from the code blocks inside the chapters.
    - Some validation.
    - Code that will replace the code block with the rendered output.
    - Code for handling the summary chapter.
- All nonterminals are now linked to the rule definition.
- The text may now link to grammar rules by just putting them in brackets like `[FunctionParameters]`. Link definitions are automatically added to every page.
- Some rules were added or changed to accommodate the new renderer. I think all changes are put into separate commits to help with reviewing.
- Various misc fixes, see the individual commits.

I'd like to thank @lukaslueg for creating the railroad library which made this possible.

Closes https://github.com/rust-lang/reference/issues/221
Closes https://github.com/rust-lang/reference/issues/398
Closes https://github.com/rust-lang/reference/issues/596
Closes https://github.com/rust-lang/reference/issues/1513
Closes https://github.com/rust-lang/reference/issues/1677

ehuss committed:
Add some missing Syntax markers

ehuss committed:
Add missing rules for syntax blocks

ehuss committed:
Fix some minor grammar formatting issues

Just fixing some small consistency and spacing mistakes.

ehuss committed:
Fix CfgAttribute name

This rule was misnamed, colliding with the existing CfgAttrAttribute.

ehuss committed:
Rename IsolatedCR to CR

This renames IsolatedCR to CR. I felt like it wasn't exactly necessary
since we have rewritten things so that it is clear that there is an
input transformation which resolves this (`input.crlf`). We also never
really defined what it meant.

I also felt like there was room for confusion. For example, an input
containing `CR CR LF LF` would get normalized to `CR LF`. The `CR` there
is not isolated.

ehuss committed:
Name common ascii control characters

This removes all backslash escaped characters. This helps to avoid
confusing similarities with a literal backslash followed by a character
versus the interpreted escaped character.

ehuss committed:
Remove "followed by" in STRING_CONTINUE

I don't exactly know why this was placed there, but we operate under the
assumption that all lexical characters immediately follow one another.

ehuss committed:
Introduce a new "prose" terminal

This introduces a new terminal kind that I'm calling a "prose" which
describes what the terminal is. This is inspired by the IETF format
which uses angle brackets to describe terminals in English.

ehuss committed:
Normalize suffix capitalization

The grammar almost always uses lowercase, so let's standardize on that.

ehuss committed:
Remove parentheses around suffixes

This helps to standardize how suffixes are written. Normally they do not
use parentheses, and visually I don't think they entirely necessary.

ehuss committed:
Fix nonterminals of ConstParam

These two nonterminals were using the wrong name for the productions for
BlockExpression and LiteralExpression.

ehuss committed:
Rewrite how keywords are listed

This changes the keyword listings so that they are just lists instead of
lexer rules. We never used the named rules, and I don't foresee us ever
doing that. Also, the `IDENTIFIER_OR_KEYWORD` rule meant that we never
needed to explicitly identify these keywords as lexer tokens.

This helps avoid problems when building the grammar graph for missing
connections.

ehuss committed:
Fix `dyn` edition presentation

Per our style, edition differences are supposed to be separated out into
an edition block.

ehuss committed:
Add grammar rule for XID_Start and XID_Continue

These were defined in prose below, but defining them here allows us to
easily refer and link to them.

ehuss committed:
Add a lexer rule for punctuation

This is intended to help define what a "token" is via the grammar (and
to fill a missing hole in our token definition).

I waffled on how to define delimiters, whether they should be separate
somehow. In practice I think it should be fine to clump them all
together. This mainly only matters for TokenTree which already excludes
the delimiters.

ehuss committed:
Add a grammar rule for reserved tokens

This adds a grammar rule that collects all the reserved token forms into
a single production rule so that we can define what a "token" is by
referring to this.

ehuss committed:
Define the Token rule

This defines a Token in the grammar so that we can easily refer to it
(and to make it easier to see what all the tokens are).

ehuss committed:
Remove escape rule

We no longer represent characters via escape sequences. These can be
confused with the literal two bytes of backslash followed by a
character. See the "common productions" list for how these are now
referred to.

ehuss committed:
Introduce a new grammar renderer

This adds an extension to mdbook-spec that will parse code-blocks in a
BNF-style grammar into a rendered format, in both markdown or as
railroad diagrams.

ehuss committed:
Add the javascript hooks for handling the new railroad grammar

This adds the hooks to toggle the visibility of the railroad grammar.
The status is stored in localstorage to keep it sticky.

ehuss committed:
Add styling for the new grammar and railroad diagrams

ehuss committed:
Add a summary chapter that shows all of the grammar productions on one page

ehuss committed:
Add some documentation for how to write grammar rules

ehuss committed:
Fix rule reference links with multiple spaces

This fixes it so that rule links work correctly if there is more than
one space in a reference definition.

ehuss committed:
Add attributes base rule.

Not sure how this got missed.

ehuss committed:
Convert all of the grammar to use the new grammar renderer

ehuss committed:
Update the introduction about grammar notation

I didn't want to try to add an unused grammar rule here, so just point
to the notation chapter which has an example.

ehuss committed:
Update the notation chapter for grammar updates

This updates some of the description for the new grammar renderer.

lukaslueg commented:
`railroad` upstream here.

The `railroad` codebase hasn't seen a lot of love with respect to graphical layout. Suggestions are welcome.

AFAICS there are cases where the grammar diverges from its graphical representation with respect to repeated elements. In the two examples below, the diagram only allows for for at least **two** consecutive `Statement` (... two consecutive `Expression`), while the grammar requires "at least one".

![Bildschirmfoto 2025-04-11 um 14 15 55](https://github.com/user-attachments/assets/fedb312f-9da3-4f31-8a6f-042b1ebe896b)
![Bildschirmfoto 2025-04-11 um 14 16 43](https://github.com/user-attachments/assets/68a7c756-c366-460e-9561-5b11c9f43a21)

ia0 commented:
> The `railroad` codebase hasn't seen a lot of love with respect to graphical layout. Suggestions are welcome.

From the live demo I can see that `*`-repeated elements are in theory printed like this:

```
   .------->-------.
   |               |
->-+-+--[ foo ]--+-+--->-
     |           |
     '-----<-----'
```

What about doing it like this?

```
->-+------>------+->-
   |             |
   '-<-[ foo ]-<-'
```

This uses one less path and can be concatenated with a previous `foo` for `+`-repeat:

```
->-[ foo ]->-+------>------+->-
             |             |
             '-<-[ foo ]-<-'
```

The main problem is that `foo` is somehow to be read backwards, which may confuse people at first.

lukaslueg commented:
> > The `railroad` codebase hasn't seen a lot of love with respect to graphical layout. Suggestions are welcome.
> 
> From the live demo I can see that `*`-repeated elements are in theory printed like this:
> 
> ```
>    .------->-------.
>    |               |
> ->-+-+--[ foo ]--+-+--->-
>      |           |
>      '-----<-----'
> ```
> 
> What about doing it like this?
> 
> ```
> ->-+------>------+->-
>    |             |
>    '-<-[ foo ]-<-'
> ```
> 
> This uses one less path and can be concatenated with a previous `foo` for `+`-repeat:
> 
> ```
> ->-[ foo ]->-+------>------+->-
>              |             |
>              '-<-[ foo ]-<-'
> ```
> 
> The main problem is that `foo` is somehow to be read backwards, which may confuse people at first.

With respect to `*`-Elements, both examples (1 and 2) are technically valid. See the [`Zero or more table-constraints` block](https://github.com/lukaslueg/railroad_dsl) in the [`create-table-stmt` example](https://github.com/lukaslueg/railroad_dsl), which demonstrates the second case.
Also see the `One or more column-definitions`-example, which should cover the `+`-Element case in example 3.

It's possible to implement `dyn Node` downstream to build more specialized primitives for certain situations. For instance, one might want to cook up a graphical representation for the "any character except ..."-case. Upstream might also provide them, if the need arises.

ia0 commented:
I see, so that's already supported and just a matter of generating the proper diagram downstream.

ehuss commented:
Oops, thanks! I messed up all the repeats.

I'm a bit uncertain on how to handle the non-greedy ones, but I think a label works ok for now.  Same with the `a..b` repeat range, and I went with a comment for that, which I think works ok?

lukaslueg commented:
Some comments and suggestions, which you hopefully find constructive

* A `LabeledBox` can use any primitive as the label, including a `Sequence`, just just plain text. See example `12_select-stmt_diagram.txt` in [railroad_dsl](https://github.com/lukaslueg/railroad_dsl). This might be useful in situations where a `Terminal` or `NonTerminal` is referenced, as in "except `\0` or `\x00`" and "except `$` and delimiters" (where "delimiters" is a `NonTerminal`); yes, the `NonTerminal` in the `Comment` can have a clickable `Link` :-) `railroad_dsl`'s syntax allows to access that, and this here might as well.
  <img width="281" alt="Bildschirmfoto 2025-04-11 um 18 22 41" src="https://github.com/user-attachments/assets/0f4ce4bd-0949-4abf-905c-694fae583101" />
  <img width="313" alt="Bildschirmfoto 2025-04-11 um 18 24 17" src="https://github.com/user-attachments/assets/2ce929a4-f962-4d20-a55a-37de4bc54020" />
* Common pre- and suffixes in diverging arms (`Choice`) can be combined, shortening the `Choice`. One can use mathematical rigor to compute that, yet I found that real-world scenarios benefit from less strictness; there is a rather ridiculously complex implementation for this in `macro_railroad` [here](https://github.com/lukaslueg/macro_railroad/blob/191c516fd2548b4a7cee21f0b99ad2ad043b11a9/src/lowering.rs#L544) (unless you can control this manually here?!).
  <img width="624" alt="Bildschirmfoto 2025-04-11 um 18 28 25" src="https://github.com/user-attachments/assets/fd2cf4d8-d04a-4ed0-9b1a-36f13014ca78" />
* There are some mistakes with respect to repeating elements in `b2c1d6ada`. Notice that reading direction reverses for the combining element (right to left, since we are going backwards within the syntax). For instance, the syntax for `StructFields` allows for "`StructField` `StructField` `,` `,`" (`struct foobar { foo: i32 bar: i32, , }`). Also notice that `WhereClause` allows for `where , T: std::fmt::Debug U: std::fmt::Debug` (both commas wrong). Unless you already did, you might want to have a look at the examples in `railroad_dsl`, which (afaics) cover those cases here (`*`/`+`/`?`).
  <img width="492" alt="Bildschirmfoto 2025-04-11 um 18 30 45" src="https://github.com/user-attachments/assets/8695d80f-f883-45f1-8af8-5bda7eff7b15" />
  <img width="571" alt="Bildschirmfoto 2025-04-11 um 18 32 37" src="https://github.com/user-attachments/assets/c6a1e975-45fe-40a2-8671-d9a7950aec93" />
* One can crate a `dyn Node` that wraps a `LabeledBox`, where the wrapping type pre-sets the label and adds another CSS-class in order to visualize `non greedy regions` (think "green box")
* With respect to "Anything except" (@traviscross) my first inclination would be a `LabeledBox` with custom CSS (think "red box"). Remember that the label can be a `dyn Node` itself, so the exceptions to the rule can at least have some unicode label (⚠) automatically attached to them.
* "Anything except" is actually just a case of "early parse error". There is no primitive for that, as we would also have to disambiguate diverging arms if anything in the diagram could lead to a parse error. AFAICS this is not a good solution.

lukaslueg commented:
The problem with "except" is that its not immediately clear if the except-element is an immediate parse error or a delayed parse error, in case both are possible. E.g. if `Foo X` is invalid but `Foo XY` is not, the graphical (stateless!) representation becomes murky and hard to follow. As suggested above, imho a comment with `NonTerminal`/`Terminal` inside them - spiced up with some css - probably suffices, as there probably are cases which are either complex or verbose, and require clear text anyway ("except `\u{0}`, `\u{00}`, …, `\u{000000}`").

These should definitely get flipped around: Whatever "a character" is should be a `Terminal`, and the exception (`LF`) should be part of the comment. If we had a `dyn Node` for "except" ("yellow box" - red is error) it would clearly indicate that there is an exception to what the `NonTerminal` can match.
 
<img width="494" alt="Bildschirmfoto 2025-04-11 um 19 46 23" src="https://github.com/user-attachments/assets/aa16a215-317d-445c-b744-7f72897b712d" />

lukaslueg commented:
See [here](https://github.com/lukaslueg/railroad_dsl/compare/master...except) for a quick and dirty example of how an `except` node might look like

<img width="459" alt="Bildschirmfoto 2025-04-11 um 21 00 53" src="https://github.com/user-attachments/assets/dfc56e8c-c850-4ebd-8936-d96e3a3718d6" />

ehuss committed:
Don't index the grammar summary appendix

This just ends up with lots of duplicates in the search results which I
don't find particularly helpful.

ehuss committed:
Reverse the Repeat expressions

Railroad renders these in reverse order, but our grammar isn't written
that way.

ehuss committed:
Use a new technique for rendering negative expressions

This is the suggestions from lukaslueg

ehuss commented:
Thanks @lukaslueg! I went ahead and added your suggestion for the Except clauses.

To fix the repeat expressions, I had to essentially reverse the Sequence blocks inside a repeat. I couldn't find any other way to get around the HDir invert.

As for some of the other suggestions here, such as rendering nodes inside comments, those sound like good things to try in the future. That in particular might be challenging because the content is written in markdown.

traviscross committed:
Mark grammar roots inline in productions

We check that the list of grammar "roots" -- that is, productions that
are not used in any other production -- is what we expect it to be.

We had hard coded this list of roots in `mdbook-spec`.  Let's instead
add a way to specify this in our syntax for productions by prefixing
the production with `@root`.

lukaslueg commented:
> The `UNICODE_ESCAPE` one is actually the correct grammar. E.g.:

👀 I was actually to lazy to check, sorry for the confusion. [The syntax is somewhat hilarious?!]

> (On `INNER_LINE_DOC`, were you meaning to point out some problem by highlighting it?)

On `INNER_LINE_DOC`, I was highlighting the fact that reading direction - graphically indicated by the arrows - is correct (green), while in `INNER_BLOCK_DOC` reading direction gets inverted on char `[CHAR]`-branch (red); as a mental image: "two trains would collide head-on in the red sections".

traviscross committed:
Cross link between grammar and railroad diagrams

When reviewing a production in the grammar, one often wants to quickly
find the corresponding railroad diagram, and when reviewing a railroad
diagram, one often wants to quickly find the corresponding production
in the grammar.

Let's make this easy by linking each production in the grammar to the
corresponding railroad diagram, and from the name of each railroad
diagram to the corresponding production in the grammar.

When clicking on a production in the grammar, we'll automatically
display the railroad diagrams if those are not already displayed.

traviscross committed:
Render `*` repeats in forward direction

There are two ways to render a "zero or more" (i.e. `*`) repeat.  One
is to put nothing on the main forward line and to put the pattern on
the recurrent edge, and the other is to put the pattern on the main
forward line and to have an empty recurrent edge and an empty bypass
edge.

That is, for the latter, we can think of `thing*` as `(thing+)?`.

Doing it that latter way means an additional edge, but it buys us
something big in return, which is that it keeps all the patterns going
in the forward direction.  Doing it the other way means the patterns
have to be reversed so as to put them underneath on that recurrent
edge, and it means that readers then have to read them right to left.

Reversing the elements also causes a bug in some diagrams where the
lines end up running in opposing directions and so the trains crash
into each other.  See:

- https://github.com/rust-lang/reference/pull/1787#issuecomment-2801343832

Keeping things in the forward direction avoids this problem.

In this commit, we'll leave in place all the infrastructure for
reversing the elements though it is no longer used.  We can of course
pull this out later.

traviscross committed:
Fix check for new and unexpected roots

We track the "roots" in our grammar -- those productions that aren't
used in any other production.  We want to report when a new root
appears or when something that's expected to be a root no longer is
one.  However, we were reporting the latter case as the former instead
of reporting it separately as intended.  Let's fix that.

ehuss commented on mdbook-spec/src/grammar/render_railroad.rs:
```diff
@@ -0,0 +1,283 @@
+        reverse: bool,
+    ) -> Option<Box<dyn Node>> {
+        let n: Box<dyn Node> = match &self.kind {
+            ExpressionKind::Grouped(e) => {
+                // I don't think this needs anything special. The grouped
+                // expression is usually an Alt or Optional or something like
+                // that which ends up as a distinct railroad node. But I'm not
+                // sure.
```
I don't know how the markdown renderer would be able to know where the groups should appear. That seems like it would be relatively complex.

ehuss commented on mdbook-spec/src/grammar.rs:
```diff
@@ -0,0 +1,389 @@
+        "OUTER_LINE_DOC",
+    ]
+    .into_iter()
+    .collect();
```
That seems fine to me.

ehuss commented on mdbook-spec/src/grammar/parser.rs:
```diff
@@ -0,0 +1,442 @@
+        let Some(next) = self.peek() else {
+            bail!(self, "expected expression after ~");
+        };
+        let kind = match next {
+            b'[' => self.parse_charset()?,
+            b'`' => self.parse_terminal()?,
+            _ => self.parse_nonterminal().ok_or_else(|| {
+                self.error("expected a charset, terminal, or name after ~ negation".to_string())
+            })?,
```
Yea, I don't like having those, either.

I would also like to get rid of the multi-character negations like `~[`*/` CR]`.  That doesn't really follow how people think of sets of characters. I'm just not sure how to encode a negative lookahead like that. I was wondering if maybe non-greedy repetition would help, but couldn't convince myself it would be equivalent.

ehuss commented on src/comments.md:
```diff
@@ -1,37 +1,40 @@
+    `//!` ~[LF CR]*
+
+INNER_BLOCK_DOC ->
+    `/*!` ( BlockCommentOrDoc | ~[`*/` CR] )* `*/`
```
I would also like to get rid of the multi-character negations like `~[`*/` CR]`.  That doesn't really follow how people think of sets of characters. I'm just not sure how to encode a negative lookahead like that. I was wondering if maybe non-greedy repetition would help, but couldn't convince myself it would be equivalent.

traviscross committed:
Remove support for reversing railroad elements

We no longer need to reverse the elements anywhere in our railroad
diagrams, so let's remove the supporting infrastructure for doing
this.

ehuss commented on mdbook-spec/src/grammar/render_railroad.rs:
```diff
@@ -0,0 +1,283 @@
+            ExpressionKind::RepeatRange(e, a, b) => {
+                let n = e.render_railroad(stack, link_map, reverse)?;
+                let cmt = match (a, b) {
+                    (Some(a), Some(b)) => format!("repeat between {a} and {b} times"),
+                    (None, Some(b)) => format!("repeat at most {b} times"),
+                    (Some(a), None) => format!("repeat at least {a} times"),
```
Yea, I agree it could be a little confusing. Open to suggestions on how to better encode that.

We could have something like `2..` show up as `(A  A+)`
And `{1..6}` to be something like `A (A (A (A (A A?)?)?)?)?`, but I'm not sure how ugly that would render.

ehuss replied:
Works for me.

ehuss commented on mdbook-spec/src/grammar/render_markdown.rs:
```diff
@@ -0,0 +1,228 @@
+                .unwrap();
+            }
+            ExpressionKind::Nt(nt) => {
+                let dest = link_map.get(nt).map_or("missing", |d| d.as_str());
```
Yes, if you have a mistake in the grammar, like you refer to a nonterminal that is not defined. This will generate a warning in the console, but I want it to still render *something*. It could render without a link, but that would add more lines of code here.

traviscross committed:
Replace a `match` with `let-else`

We can save a line by replacing this `match` with a `let-else`, so
let's do that.

ehuss commented on mdbook-spec/src/grammar/render_railroad.rs:
```diff
@@ -0,0 +1,283 @@
+        let seq: Sequence<Box<dyn Node>> =
+            Sequence::new(vec![Box::new(SimpleStart), n.unwrap(), Box::new(SimpleEnd)]);
+        let vert = VerticalGrid::<Box<dyn Node>>::new(vec![
+            Box::new(Comment::new(self.name.clone())),
```
Sounds good!

ehuss commented on mdbook-spec/src/grammar.rs:
```diff
@@ -0,0 +1,389 @@
+    .collect();
+    if set != expected {
+        let new: Vec<_> = set.symmetric_difference(&expected).collect();
+        let removed: Vec<_> = expected.symmetric_difference(&set).collect();
```
Oops, yea, just picked the wrong function.

traviscross committed:
Lower `RepeatRange` to railroads more elaborately

For `RepeatRange(e, a, b)`, we were rendering `e` on the main line
then rendering under it a message about how many times it may or must
repeat based on `a` and `b`.

The trouble is that if we say that something "repeats once" on the
recurrent edge -- after we've already consumed a thing -- that reads
reasonably as though we're saying that two things can be consumed when
that's not what we mean.

Similarly, it's a bit odd to say, on the recurrent edge, that
something must "repeat twice" when that edge (and presumably then that
rule) may not be taken at all.

Let's solve all this by doing the following:

- For `e{1..1}`, simply render the node.
- For `e{0..1}`, treat this as simply `e?`.
- For `e{0..}`, treat this as simply `e*`.
- For `e{1..}`, treat this as simply `e+`.
- For `e{a..0}`, render an empty node.
- For `e{0..b} b > 1`, treat this as `(e{1..b})?`.
- For `e{1..b} b > 1`, render the node on the main line, then on the
  recurrent line render "at most {b - 1} more times".
- For `e{a..b} a > 1`, make a sequence of length `a` where the final
  node repeats `{1..b - (a - 1)}` times (or `{1..}` times if `b` is
  unbounded).

(We'll also add a check in parsing to ensure that for the range to be
well formed `a <= b`.)

As it turns out, the most straightforward way to implement this isn't
by recursing.  Doing that means we end up needing to take special care
to handle the suffix and the footnote, we have to build up an extra
`Expression` we don't need, and we have to `unwrap` the call.
Instead, it works better to treat this lowering in the manner of a
transitioning state machine in the spirit of `loop match` as proposed
in RFC 3720.
